src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixPkcs12Reader.cs

```diff
@@ -0,0 +1,670 @@
+                        default:
+                            return false;
+                    }
+
+                    return
+                        publicKeyInfo.Algorithm.Parameters.HasValue &&
+                        publicKeyInfo.Algorithm.Parameters.Value.Span.SequenceEqual(keyParams);
+            }
+
+            if (algorithm != publicKeyInfo.Algorithm.Algorithm.Value)
+            {
+                return false;
+            }
+
+            if (!publicKeyInfo.Algorithm.Parameters.HasValue)
+            {
+                return (keyParams?.Length ?? 0) == 0;
+            }
+
+            return publicKeyInfo.Algorithm.Parameters.Value.Span.SequenceEqual(keyParams);
+        }
+
+        private static int FindMatchingKey(
+            SafeBagAsn[] keyBags,
+            int keyBagCount,
+            ReadOnlySpan<byte> localKeyId)
+        {
+            for (int i = 0; i < keyBagCount; i++)
+            {
+                foreach (AttributeAsn attr in keyBags[i].BagAttributes)
+                {
+                    if (attr.AttrType.Value == Oids.LocalKeyId && attr.AttrValues.Length > 0)
+                    {
+                        ReadOnlyMemory<byte> curKeyId =
+                            Helpers.DecodeOctetStringAsMemory(attr.AttrValues[0]);
+
+                        if (curKeyId.Span.SequenceEqual(localKeyId))
+                        {
+                            return i;
+                        }
+
+                        break;
+                    }
+                }
+            }
+
+            return -1;
+        }
+
+        private static void DecryptSafeContents(
+            ReadOnlySpan<char> password,
+            ref ContentInfoAsn safeContentsAsn)
+        {
+            EncryptedDataAsn encryptedData =
+                EncryptedDataAsn.Decode(safeContentsAsn.Content, AsnEncodingRules.BER);
+
+            // https://tools.ietf.org/html/rfc5652#section-8
+            if (encryptedData.Version != 0 && encryptedData.Version != 2)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            // Since the contents are supposed to be the BER-encoding of an instance of
+            // SafeContents (https://tools.ietf.org/html/rfc7292#section-4.1) that implies the
+            // content type is simply "data", and that content is present.
+            if (encryptedData.EncryptedContentInfo.ContentType != Oids.Pkcs7Data)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            if (!encryptedData.EncryptedContentInfo.EncryptedContent.HasValue)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            int encryptedValueLength = encryptedData.EncryptedContentInfo.EncryptedContent.Value.Length;
+
+            byte[] destination = CryptoPool.Rent(encryptedValueLength);
```

Comment:
how is this being returned?

Reply:
The sentinel on the ContentType value tells Dispose that the Content was actually a rented segment to return.  So, Dispose.

But this was a really, really good question... I added some analysis in CryptoPool.cs and found that I had several hundred double-returns, then it swung to several hundred un-returned, then I fixed the Dispose loop to use the `i` value instead of always using `0`.